engine/src/main/java/io/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java

```diff
@@ -245,18 +242,26 @@ public void takeSequenceFlow(
         context.copy(
             sequenceFlowKey, followUpInstanceRecord, ProcessInstanceIntent.SEQUENCE_FLOW_TAKEN);
 
-    final boolean shouldActivateTargetElement;
-    if (target.getElementType() != BpmnElementType.PARALLEL_GATEWAY) {
-      shouldActivateTargetElement = true;
-    } else {
-      // if all incoming sequence flows are taken at least once, the gateway can be activated
-      final var tokensBySequenceFlow =
-          sequenceFlowAnalyzer.determineTakenIncomingFlows(context.getFlowScopeKey(), target);
-      shouldActivateTargetElement = tokensBySequenceFlow.size() == target.getIncoming().size();
+    if (canActivateTargetElement(context, target)) {
+      activateElementInstanceInFlowScope(sequenceFlowTaken, target);
     }
+  }
 
-    if (shouldActivateTargetElement) {
-      activateElementInstanceInFlowScope(sequenceFlowTaken, target);
+  private boolean canActivateTargetElement(
+      final BpmnElementContext context, final ExecutableFlowNode targetElement) {
+
+    final int numberOfIncomingSequenceFlows = targetElement.getIncoming().size();
+
+    if (targetElement.getElementType() == BpmnElementType.PARALLEL_GATEWAY
+        && numberOfIncomingSequenceFlows > 1) {
```

Comment:
Nitpicky comment: I think it is possible to have a joining parallel gateway with only one incoming sequence flow. (at least I can model it, not sure if deploy validation would refuse it)

This would not enter this clause, but the next one. So the end effect is the same. But to make the code a little clearer, I don't think the condition `numberOfIncomingSequenceFlows > 1` is necessary.

Reply:
Yes, the check of the number of sequence flows is not necessary. It is an optimization. 
Thinking about it again, we don't need it because it is uncommon to have only one incoming flow.